enhancements/fbc-run-bundle.md

```diff
@@ -0,0 +1,268 @@
+        - If the bundle is not present in the index image, Operator SDK will add the contents of the bundle image to the existing index image and serve the resulting index image (with the bundle added to it).
+
+    - Note: If the index image that was passed is in SQLite format, Operator SDK will use a different container creation command that uses `opm registry add` to add the bundle to the index and `opm registry serve` to serve the index which is a database file on created on the pod.
+
+5. The Operator SDK will invoke more Operator Registry APIs to ensure that the FBC that was generated from the above steps is indeed a valid FBC. This steps essentially validates the generated FBC by converting the declarative config object to a model object which internally calls the API-equivalent of the `opm validate <FBC>` command.
+
+6. Three unique data points will be stored in the `IndexImageCatalogCreator` object that will eventually be used in the ephemeral pod:
+
+    - FBC directory name where the FBC file resides in → `FBCDir`
+    - FBC file name that has all the contents to be served through the pod → `FBCFile`
+    - Contents of file-based catalog file (string YAML/JSON) → `FBCContent`
+
+7. During the creation of the ephemeral pod (which is used to serve the file-based catalog to gRPC), the Operator SDK will invoke a set of CLI commands to create a directory and file on the pod during runtime.
+
+8. The contents of the generated File-Based Catalog will be written to the file that is created on the ephemeral pod using the following command template -
+
+    ```go
+    `mkdir -p {{ .FBCDir }} && \
+    echo '{{ .FBCContent }}' >> {{ .FBCFile }} && \
+    opm serve {{ .FBCDir }} -p {{ .GRPCPort }}`
+    ```
+
+    - Note: It is worth noting that the new command to serve a file-based catalog to GRPC has been changed from `opm registry serve` to `opm serve`. The above command template reflects this change. `opm registry` commands will be deprecated in a future release.
+
+9. In this way, the file-based catalog that was generated by the Operator SDK will be served to the gRPC API via the specified port (defaulted to `50051`) without any user-facing changes.
+
+10. The remaining resources (`CatalogSource`, `Subscription`, `OperatorGroup`) are created in the same way as before. The run bundle command does not change from here on.
+
+- Note: If a SQLite index is passed, the old commands will be used in the registry pod and none of the FBC creation will happen. Below is the container creation command used to handle SQLite index images (the optional flags are skipped here for simplicity purposes):
+
+    ```go
+    `mkdir -p {{ dirname .DBPath }} && \
+    {{- range $i, $item := .BundleItems }}
+    opm registry add -d {{ $.DBPath }} -b {{ $item.ImageTag }} --mode={{ $item.AddMode }} && \
+    {{- end }}
+    opm registry serve -d {{ .DBPath }} -p {{ .GRPCPort }}`
+    ```
```

Comment:
IHMO if the is in SQL format then we should call opm render OR opm migrate to convert to FBC and work with.
So that, we do not have not any additional implementation to keep maintained

Reply:
@camilamacedo86 

We initially took that approach of converting the SQLite images to FBC. However, we are unable to use `opm render` or `opm migrate` to convert the SQLite images to FBC format. There are a bunch of compilation errors when trying to convert SQLite images to FBC format. 

Let's say a user creates a SQLite index image, they will need to create the image with the right build flags. If not, `run bundle` will fail trying to render that image using Operator Registry APIs. Hence, we decided to stick to the old commands such as `opm registry add` and `opm registry serve` to avoid all of the CGO and SQLite compilation issues.

Reply:
To state an example, we tried to create a SQLite based image from the [OpenShift doc](https://docs.openshift.com/container-platform/4.9/operators/admin/olm-managing-custom-catalogs.html#olm-updating-index-image_olm-managing-custom-catalogs) and tried to render it. 

Here are the errors we got: 
`Binary was compiled with 'CGO_ENABLED=0', go-sqlite3 requires cgo to work`. To get rid of this error, we built the image by changing CGO_ENABLED=0 to CGO_ENABLED=1 in our Makefile and we were able to get past the error. 

This was the next error that came up when trying to render the image `populate channels: no such function: json_object`. To get past this, I think we need to build the image using `-tags json1` in the go build invocation. 

And I think building with CGO has these complications and it's a pain to get it to cross-compile. 

From a user experience, it would be bad for `run bundle` to fail with these errors when a user provides SQLite images. We we will be much better off by navigating through `opm registry add` command instead of trying to do `opm render` for SQLite images. 

I think we have consensus with @joelanford as well on this opinion.

Reply:
I would like to check @joelanford thoughts here.
Because it does not make sense to me.

We are actually using OPM render | OPM migrate in the pipelines and with the downstream images.
Also, did we check the error faced? what occurs? when it occurs? 
is it a really expected error or is it a bug? Did we do something wrong in the tests to generate the SQL index to do the tests?

Reply:
IMO, its better to just keep the deprecated code paths the same and separate from the FBC code paths. That way, when sqlite support is dropped, we can just drop those code paths.

I don't think it makes sense in this case to "opm migrate" an sqlite catalog to an FBC since:
1. As Camila said, they should be equivalent, so there's no functional difference
2. It takes more time to actually do that conversion, so there would be a degredation to the UX if we suddenly started performing extra work for sqlite indexes.
3. The compilation problems that Rashmi mentioned are almost a complete show-stopper in and of themselves. It took a _ton_ of effort to get the `opm` binary and repo to a point that we could cross-compile it for all the target OSes and arches. We'd have to repeat that effort (and then maintain all of that extra build complexity) for SDK.

You might think we could avoid problem 3 by adding the extra step to the pod entrypoint to do `opm migrate`. However, not all sqlite indexes will contain an `opm` binary that has the `migrate` command, so we'd be breaking `run bundle` in those cases.

Reply:
Thank you for the input.